### PR TITLE
fix: [generate_interactions] per interleaved model state

### PR DESCRIPTION
**Pull request by lalo**

### CODE REVIEW

File: vowpalwabbit/core/src/reductions/generate_interactions.cc [hunk 1]
```diff
@@ -23,8 +23,9 @@ using namespace VW::config;
 namespace
 {
 template <bool is_learn, VW::generate_func_t<VW::namespace_index> generate_func, bool leave_duplicate_interactions>
-void transform_single_ex(VW::interactions_generator& data, VW::LEARNER::learner& base, VW::example& ec)
+void transform_single_ex(std::vector<VW::interactions_generator>& datas, VW::LEARNER::learner& base, VW::example& ec)
 {
+  auto data = datas[ec.ft_offset / 4];  // hard coded 4 should be _all->weights.stride();
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/reductions/generate_interactions.cc:28]: Or more generally, the increment of the below stack?

File: vowpalwabbit/core/src/reductions/generate_interactions.cc [hunk 2]
```diff
@@ -45,8 +46,9 @@ void transform_single_ex(VW::interactions_generator& data, VW::LEARNER::learner&
 
 template <bool is_learn, VW::generate_func_t<VW::namespace_index> generate_func,
     VW::generate_func_t<VW::extent_term> generate_func_extents, bool leave_duplicate_interactions>
-void transform_single_ex(VW::interactions_generator& data, VW::LEARNER::learner& base, VW::example& ec)
+void transform_single_ex(std::vector<VW::interactions_generator>& datas, VW::LEARNER::learner& base, VW::example& ec)
 {
+  auto data = datas[ec.ft_offset / 4];
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/reductions/generate_interactions.cc:51]: Make sure to reference not copy

File: vowpalwabbit/core/src/reductions/generate_interactions.cc [hunk 3]
```diff
@@ -206,6 +206,8 @@ std::shared_ptr<VW::LEARNER::learner> VW::reductions::generate_interactions_setu
     }
   }
 
+  if (options.was_supplied("automl")) { interactions_spec_contains_wildcards = true; }
```
Comment (lalo) [inline vowpalwabbit/core/src/reductions/generate_interactions.cc:209]: this in theory should be per ft_offset - not all of them have wildcards

thoughts @bassmang

Comment (lalo) [reply vowpalwabbit/core/src/reductions/generate_interactions.cc:209]: with the added challenge that the configs change ever so often

File: vowpalwabbit/core/include/vw/core/interactions.h [hunk 4]
```diff
@@ -437,7 +465,45 @@ class interactions_generator
...
+      offset_2 /= _stride;
+    }
+
+    std::swap(per_model_interactions[offset_1], per_model_interactions[offset_2]);
+    std::swap(per_model_extent_interactions[offset_1], per_model_extent_interactions[offset_2]);
+    std::swap(_count_when_last_generated[offset_1], _count_when_last_generated[offset_2]);
+  }
+
 private:
+  std::vector<gen_interactions_vec> per_model_interactions;
```
Comment (olgavrou) [inline vowpalwabbit/core/include/vw/core/interactions.h:503]:
```suggestion
  std::vector<gen_interactions_vec> _per_model_interactions;
```

File: vowpalwabbit/core/include/vw/core/interactions.h [hunk 5]
```diff
@@ -437,7 +465,45 @@ class interactions_generator
...
+  std::vector<gen_extent_interactions_vec> per_model_extent_interactions;
```
Comment (olgavrou) [inline vowpalwabbit/core/include/vw/core/interactions.h:504]:
```suggestion
  std::vector<gen_extent_interactions_vec> _per_model_extent_interactions;
```

File: vowpalwabbit/core/include/vw/core/automl_impl.h [hunk 6]
```diff
@@ -217,6 +218,7 @@ class interaction_config_manager
   bool reward_as_cost;
   double tol_x;
   bool is_brentq;
+  VW::interactions_generator _interactions_cache;
```
Comment (olgavrou) [inline vowpalwabbit/core/include/vw/core/automl_impl.h:221]: this is public, should it be without the underscore in the name or made private?

Comment (lalo) [reply vowpalwabbit/core/include/vw/core/automl_impl.h:221]: automl is in a weird state where we need a lot of things to be public just to facilitate asserting on internal state on the unit tests, to try to keep control of edge cases and specific states we now should be true after X events.

I haven't figured out a better way to allow the unit test to have full access.

File: vowpalwabbit/core/src/reductions/automl.cc [hunk 7]
```diff
@@ -168,10 +168,14 @@ std::shared_ptr<VW::LEARNER::learner> make_automl_with_impl(VW::setup_base_i& st
     trace_file_name_prefix = oss.str();
   }
 
+  auto* interactions_cache =
```
Comment (olgavrou) [inline vowpalwabbit/core/src/reductions/automl.cc:171]: are we ever worried that this might be nullptr?

File: test/train-sets/ref/automl_test_load_override.stderr [hunk 8]
```diff
@@ -6,7 +6,7 @@ learning rate = 0.5
 initial_t = 3
 power_t = 0.5
 cb_type = mtr
-Enabled learners: gd, scorer-identity, csoaa_ldf-rank, cb_adf, automl, cb_explore_adf_greedy, shared_feature_merger
+Enabled learners: gd, generate_interactions, scorer-identity, csoaa_ldf-rank, cb_adf, automl, cb_explore_adf_greedy, shared_feature_merger
```
Comment (lalo) [inline test/train-sets/ref/automl_test_load_override.stderr:9]: this diff means that the bug wasnt affecting the usual stack without ccb
